polybench/2mm.cpp

```diff
@@ -1,176 +1,159 @@
-
-		buffer<DATA_TYPE, 2> C_buffer(range<2>(NJ, NI));
-		initDeviceBuffer(queue, C_buffer, C.data());
-
-		buffer<DATA_TYPE, 2> D_buffer(range<2>(NL, NJ));
-		initDeviceBuffer(queue, D_buffer, D.data());
-
-		buffer<DATA_TYPE, 2> E_buffer(range<2>(NL, NI));
+		buffer<DATA_TYPE, 2> A_buffer(A.data(), range<2>(size, size));
+		buffer<DATA_TYPE, 2> B_buffer(B.data(), range<2>(size, size));
+		buffer<DATA_TYPE, 2> C_buffer(C.data(), range<2>(size, size));
+		buffer<DATA_TYPE, 2> D_buffer(D.data(), range<2>(size, size));
+		buffer<DATA_TYPE, 2> E_buffer(E.data(), range<2>(size, size));
 
-		double t_start = rtclock();
-
-		queue.submit([&](handler& cgh) {
-			auto A_access = A_buffer.get_access<access::mode::read>(cgh);
-			auto B_access = B_buffer.get_access<access::mode::read>(cgh);
-			auto C_access = C_buffer.get_access<access::mode::discard_write>(cgh);
-
-			const auto range = nd_range<2>(C_buffer.get_range(), {DIM_THREAD_BLOCK_Y, DIM_THREAD_BLOCK_X});
-
-			cgh.parallel_for<class MM1>(range, [=, NK_ = NK](nd_item<2> nd_item) {
-				const auto item = nd_item.get_global_id();
+		args.device_queue.wait_and_throw();
+		args.device_queue.submit([&](handler& cgh) {
+			auto A = A_buffer.get_access<access::mode::read>(cgh);
+			auto B = B_buffer.get_access<access::mode::read>(cgh);
+			auto C = C_buffer.get_access<access::mode::read_write>(cgh);
 
+			cgh.parallel_for<class Polybench_2mm_1>(C_buffer.get_range(), [=, size_ = size](item<2> item) {
 				const auto i = item[0];
 				const auto j = item[1];
 
-				C_access[item] = 0;
-				for(size_t k = 0; k < NK_; k++) {
-					C_access[item] += A_access[{i, k}] * B_access[{k, j}];
+				for(size_t k = 0; k < size_; k++) {
+					C[item] += A[{i, k}] * B[{k, j}];
 				}
 			});
 		});
 
-		queue.submit([&](handler& cgh) {
-			auto C_access = C_buffer.get_access<access::mode::read>(cgh);
-			auto D_access = D_buffer.get_access<access::mode::read>(cgh);
-			auto E_access = E_buffer.get_access<access::mode::discard_write>(cgh);
-
-			const auto range = nd_range<2>(E_buffer.get_range(), {DIM_THREAD_BLOCK_Y, DIM_THREAD_BLOCK_X});
-
-			cgh.parallel_for<class MM2>(range, [=, NJ_ = NJ](nd_item<2> nd_item) {
-				const auto item = nd_item.get_global_id();
+		args.device_queue.submit([&](handler& cgh) {
+			auto C = C_buffer.get_access<access::mode::read>(cgh);
+			auto D = D_buffer.get_access<access::mode::read>(cgh);
+			auto E = E_buffer.get_access<access::mode::discard_write>(cgh);
 
+			cgh.parallel_for<class Polybench_2mm_2>(E_buffer.get_range(), [=, size_ = size](item<2> item) {
 				const auto i = item[0];
 				const auto j = item[1];
 
-				E_access[item] = 0;
-				for(size_t k = 0; k < NJ_; k++) {
-					E_access[item] += C_access[{i, k}] * D_access[{k, j}];
+				E[item] = 0;
+				for(size_t k = 0; k < size_; k++) {
+					E[item] += C[{i, k}] * D[{k, j}];
 				}
 			});
 		});
 
-		queue.wait();
-		double t_end = rtclock();
+		args.device_queue.wait_and_throw();

```

Comment:
This wait is not necessary, because the benchmarking framework already calls `wait_and_throw()` after `run()`.